### PR TITLE
ddl: Add admin check before drop table using building flag.

### PR DESCRIPTION
**Pull request by crazycs520**
## What have you changed? (mandatory)
Add `admin check table` before drop table according to the ldflags when building.

## Does this PR affect tidb-ansible update? (mandatory)
No

## Does this PR need to be added to the release notes? (mandatory)
No

### CODE REVIEW

**Comment by crazycs520** [comment]
@winkyao @zimulala @shenli PTAL

File: executor/ddl.go [hunk 1]
```diff
@@ -194,6 +197,14 @@ func (e *DDLExec) executeDropTable(s *ast.DropTableStmt) error {
 			return errors.Trace(err)
 		}
 
+		if os.Getenv("TIDB_CHECK_BEFORE_DROP") == "1" {
```
Comment (shenli) [inline executor/ddl.go]: Please add a log here. I suggest using the warn log level.

File: executor/executor.go [hunk 2]
```diff
@@ -371,6 +371,9 @@ func (e *CheckTableExec) Next(ctx context.Context, chk *chunk.Chunk) error {
 	defer func() { e.done = true }()
 	dbName := model.NewCIStr(e.ctx.GetSessionVars().CurrentDB)
 	for _, t := range e.tables {
+		if dbName.L == "" {
```
Comment (shenli) [inline executor/executor.go]: Is this a bug-fix?

File: executor/executor_test.go [hunk 3]
```diff
@@ -247,6 +252,10 @@ func (s *testSuite) TestAdmin(c *C) {
 	// For "checksum_without_index", we only have one checksum, so the result will be 1.
 	res.Sort().Check(testkit.Rows("test checksum_with_index 0 2 2", "test checksum_without_index 1 1 1"))
 
+	tk.MustExec("drop table if exists admin_test")
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/executor_test.go]: done.

File: executor/executor.go (hunk 2 above)
Comment (crazycs520) [reply executor/executor.go]: Ye~
If use `e.ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(e.ctx, "admin check table test.t1")` ,here `dbName.L` is "".

File: executor/ddl.go (hunk 1 above)
Comment (crazycs520) [reply executor/ddl.go]: done.

File: executor/executor_test.go [hunk 4]
```diff
@@ -236,6 +240,13 @@ func (s *testSuite) TestAdmin(c *C) {
 	r, err = tk.Exec("admin check table admin_test")
 	c.Assert(err, NotNil)
 
+	os.Setenv("TIDB_CHECK_BEFORE_DROP", "1")
+	r, err = tk.Exec("drop table admin_test")
```
Comment (winkyao) [inline executor/executor_test.go]: How can we know the admin check is executed?

Comment (winkyao) [reply executor/executor_test.go]: You use gofail to inject an error in executeDropTable.

Comment (crazycs520) [reply executor/executor_test.go]: `admin check table admin_test` must be error, so `drop table admin_test` must be error to know `admin check` is executed.

File: executor/executor_test.go [hunk 5]
```diff
@@ -236,6 +240,13 @@ func (s *testSuite) TestAdmin(c *C) {
...
+	c.Assert(err, NotNil)
```
Comment (shenli) [inline executor/executor_test.go]: Please check the content of the error to make sure it is thrown by the consistent checking.

**Comment by winkyao** [comment]
@crazycs520 Please add proper label.

**Comment by jackysp** [comment]
I think it's better to use a building flag for this check.

File: executor/executor.go (hunk 2 above)
Comment (zimulala) [reply executor/executor.go]: Could we use `t.DBInfo.Name` directly?

Comment (zimulala) [reply executor/executor.go]: @crazycs520

**Comment by zz-jason** [comment]
depends on os environment variable is not a good practice.

**Comment by crazycs520** [comment]
@jackysp @zz-jason @tiancaiamao PTAL

**Comment by winkyao** [comment]
@crazycs520 Please change the pr title.

File: config/config.go [hunk 6]
```diff
@@ -36,8 +36,18 @@ var (
...
 	}
+	// CheckBeforeDrop enable to execute `admin check table` before `drop table`.
+	CheckBeforeDrop = false
+	// CheckBeforeDropLDFlag is a go build flag.
+	CheckBeforeDropLDFlag = "None"
 )
 
+func init() {
+	if CheckBeforeDropLDFlag == "1" {
+		CheckBeforeDrop = true
```
Comment (winkyao) [inline config/config.go]: use atomic to protect `CheckBeforeDrop`, or the data race test perhaps reports error.

File: executor/executor_test.go [hunk 7]
```diff
@@ -130,6 +130,10 @@ func (s *testSuite) TearDownTest(c *C) {
 }
 
 func (s *testSuite) TestAdmin(c *C) {
+	origin := config.CheckBeforeDrop
+	defer func() {
+		config.CheckBeforeDrop = origin
```
Comment (winkyao) [inline executor/executor_test.go]: ditto

File: util/printer/printer.go [hunk 8]
```diff
@@ -43,6 +43,7 @@ func PrintTiDBInfo() {
 	log.Infof("UTC Build Time:  %s", TiDBBuildTS)
 	log.Infof("GoVersion:  %s", GoVersion)
 	log.Infof("Race Enabled: %v", israce.RaceEnabled)
+	log.Infof("Admin check table before drop is %v", config.CheckBeforeDrop)
```
Comment (lamxTyler) [inline util/printer/printer.go]: How about `Check Table Before Drop: %v`?

File: executor/ddl.go [hunk 9]
```diff
@@ -194,6 +198,15 @@ func (e *DDLExec) executeDropTable(s *ast.DropTableStmt) error {
 			return errors.Trace(err)
 		}
 
+		if config.CheckBeforeDrop {
+			log.Warnf("admin check table before drop was on")
```
Comment (jackysp) [inline executor/ddl.go]: It's better to add more info in this log, e.g. the table name.

File: executor/executor.go [hunk 10]
```diff
@@ -369,8 +369,8 @@ func (e *CheckTableExec) Next(ctx context.Context, chk *chunk.Chunk) error {
 		return nil
 	}
 	defer func() { e.done = true }()
-	dbName := model.NewCIStr(e.ctx.GetSessionVars().CurrentDB)
 	for _, t := range e.tables {
+		dbName := t.DBInfo.Name
```
Comment (jackysp) [inline executor/executor.go]: Finally, it fixed. 👍

File: config/config.go (hunk 6 above)
Comment (crazycs520) [reply config/config.go]: done.

File: util/printer/printer.go (hunk 8 above)
Comment (crazycs520) [reply util/printer/printer.go]: done.

File: executor/ddl.go (hunk 9 above)
Comment (crazycs520) [reply executor/ddl.go]: done.

File: executor/executor.go (hunk 10 above)
Comment (crazycs520) [reply executor/executor.go]: done.

**Comment by crazycs520** [comment]
@winkyao PTAL again. I add `ldflags` to test and remove `atomic` set.

File: config/config.go [hunk 11]
```diff
@@ -36,6 +36,10 @@ var (
 		"mocktikv": true,
 		"tikv":     true,
 	}
+	// checkTableBeforeDrop enable to execute `admin check table` before `drop table`.
+	CheckTableBeforeDrop = false
+	// CheckBeforeDropLDFlag is a go build flag.
+	CheckBeforeDropLDFlag = "None"
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply config/config.go]: done. PTAL

**Comment by zz-jason** [comment]
should we also add some UTs to make sure the `ldflags` take effect?

**Comment by tiancaiamao** [comment]
LGTM
It's hard to use unit test to check `ldflags`, we have to do it manually or in the integrated test. @zz-jason

**Comment by zz-jason** [comment]
@tiancaiamao If we run `make test`, the global config `config.CheckTableBeforeDrop` should always be true, maybe we can check this? @crazycs520 How do you think?

**Comment by zz-jason** [comment]
LGTM

**Pull request merged**